base/loading.jl

```diff
@@ -1362,20 +1350,26 @@ function include_package_for_output(pkg::PkgId, input::String, depot_path::Vecto
     Base._track_dependencies[] = true
     get!(Base.PkgOrigin, Base.pkgorigins, pkg).path = input
     append!(empty!(Base._concrete_dependencies), concrete_deps)
-    uuid_tuple = pkg.uuid === nothing ? (UInt64(0), UInt64(0)) : convert(NTuple{2, UInt64}, pkg.uuid)
-
-    ccall(:jl_set_module_uuid, Cvoid, (Any, NTuple{2, UInt64}), Base.__toplevel__, uuid_tuple)
     if source !== nothing
         task_local_storage()[:SOURCE_PATH] = source
     end
 
+    newm = Module(Symbol(pkg.name), false)
+    uuid = pkg.uuid
+    if uuid !== nothing
+        ccall(:jl_set_module_uuid, Cvoid, (Any, NTuple{2, UInt64}), newm, uuid)
+        # HACK for preferences
+        ccall(:jl_set_module_uuid, Cvoid, (Any, NTuple{2, UInt64}), Main, uuid)
+    end
+    @lock require_lock register_root_module(newm, pkg)
     try
-        Base.include(Base.__toplevel__, input)
+        eval(newm, :(baremodule $(nameof(newm)); $include($newm, $input); end))
```

Comment:
Is the extra `baremodule` layer doing anything here? It seems to me evaluating inside `newm` is already enough?

Reply:
Running `__init__` functions is tied to syntactic module definitions